python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,346 @@
+import copy
+from abc import ABCMeta, abstractmethod
+from collections import defaultdict
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Dict, List, Optional
+
+from google.protobuf.json_format import MessageToDict
+
+from ray.autoscaler._private.resource_demand_scheduler import UtilizationScore
+from ray.autoscaler.v2.schema import NodeType
+from ray.core.generated.autoscaler_pb2 import (
+    ClusterResourceConstraint,
+    GangResourceRequest,
+    NodeState,
+    ResourceRequest,
+    ResourceRequestByCount,
+)
+from ray.core.generated.instance_manager_pb2 import Instance
+
+# ============= Resource Scheduling Service API =======================
+#
+#  ResourceSchedulerService is a service that schedules resource bundles
+#  to nodes. It's used by the autoscaler to schedule resource bundles
+#  to determine the desired cluster size to satisfy the current resource
+#  demands.
+#
+
+
+@dataclass
+class NodeTypeConfig:
+    # Node type name
+    name: NodeType
+    # The minimal number of workers to be launched for this node type.
+    min_workers: int
+    # The maximal number of workers can be launched for this node type.
+    max_workers: int
+    # The resources on the node.
+    resources: Dict[str, float] = field(default_factory=dict)
+    # The labels on the node.
+    labels: Dict[str, str] = field(default_factory=dict)
+
+
+@dataclass
+class ResourceScheduleConfig:
+    # The node type configs.
+    node_type_configs: List[NodeTypeConfig] = field(default_factory=list)
+    # The max number of worker nodes to be launched for the entire cluster.
+    max_num_worker_nodes: Optional[int]
+
+
+@dataclass
+class ResourceScheduleRequest:
+    # TODO: This prob could be refactored into the ClusterStatus data class later.
+    # The current ray resource requests.
+    resource_requests: List[ResourceRequestByCount] = field(default_factory=list)
+    # The Gang resource requests.
+    gang_resource_requests: List[GangResourceRequest] = field(default_factory=list)
+    # cluster resource constraints.
+    cluster_resource_constraints: List[ClusterResourceConstraint] = field(
+        default_factory=list
+    )
+    # The ray nodes
+    node_states: List[NodeState] = field(default_factory=list)
+    # The current list of instances.
+    current_instances: List[Instance] = field(default_factory=list)
+    # The resource schedule config.
+    schedule_config: ResourceScheduleConfig
```

Comment:
We need to know 
1. the current ray cluster's load and resources uage: thus `node_states`
2. the pending instances to avoid double launch: thus `current_instances`.